### PR TITLE
add job priority for kube-batch scheduling #141

### PR DESCRIPTION
**Pull request by YesterdayxD**
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/common/45)

### CODE REVIEW

File: job_controller/job.go [hunk 1]
```diff
@@ -191,6 +191,16 @@ func (jc *JobController) ReconcileJobs(
 		return jc.Controller.UpdateJobStatusInApiServer(job, &jobStatus)
 	}
 
+	if jc.Config.EnableGangScheduling {
+		minAvailableReplicas := getTotalReplicas(replicas)
+		priorityClassName := getPriorityClassName(runPolicy)
+		//_, err := pc.SyncPodGroup(job, minAvailableReplicas)
```
Comment (merlintang) [inline job_controller/job.go]: why keep this ?

Comment (YesterdayxD) [reply job_controller/job.go]: I reference the code about EnableGangScheduling of other operators and add the priority‘s’attribute and function.

File: job_controller/api/v1/types.go [hunk 2]
```diff
@@ -188,4 +188,8 @@ type RunPolicy struct {
 // job, for example `minAvailable` for gang-scheduling.
 type SchedulingPolicy struct {
 	MinAvailable *int32 `json:"minAvailable,omitempty"`
+
+	//add PriorityClassName
```
Comment (gaocegege) [inline job_controller/api/v1/types.go]: We can remove this line

File: job_controller/job.go [hunk 3]
```diff
@@ -40,7 +40,7 @@ func (jc *JobController) deletePodsAndServices(runPolicy *apiv1.RunPolicy, job i
 	}
 	return nil
 }
-
+// TTL means time to live
```
Comment (gaocegege) [inline job_controller/job.go]: This line, too.

File: job_controller/job.go [hunk 4]
```diff
@@ -299,3 +309,14 @@ func (jc *JobController) cleanupJob(runPolicy *apiv1.RunPolicy, jobStatus apiv1.
 	jc.WorkQueue.AddRateLimited(key)
 	return nil
 }
+func getPriorityClassName(runPolicy *apiv1.RunPolicy) string {
+	priorityClassName := *runPolicy.SchedulingPolicy.PriorityClassName
```
Comment (gaocegege) [inline job_controller/job.go]: Runtime error here

File: job_controller/job.go [hunk 5]
```diff
@@ -299,3 +309,14 @@ func (jc *JobController) cleanupJob(runPolicy *apiv1.RunPolicy, jobStatus apiv1.
...
+	return priorityClassName
+}
+func getTotalReplicas(replicas map[apiv1.ReplicaType]*apiv1.ReplicaSpec) int32 {
+	jobReplicas := int32(0)
+	for _, r := range replicas {
+		jobReplicas += *r.Replicas
```
Comment (gaocegege) [inline job_controller/job.go]: I think we should check if r.Replicas is nil here, too.

Comment (YesterdayxD) [reply job_controller/job.go]: I got it

**Comment by YesterdayxD** [comment]
I try to correct it

**Comment by terrytangyuan** [comment]
Hmm I just asked the author of this PR to submit the changes here: https://github.com/kubeflow/mpi-operator/pull/141

Are you in the same team? It would be better if you can send a note before submitting the PR to avoid redundant work...

**Comment by terrytangyuan** [comment]
cc @4everming

**Comment by YesterdayxD** [comment]
@terrytangyuan Yes, we are in same team. I will be more careful next time. Thank you for reminding me.
 @4everming

**Comment by gaocegege** [comment]
@terrytangyuan Since the MPI operator does not use the common operator now, should we keep the similar logic in these two repos?

**Comment by terrytangyuan** [comment]
Yes I think it’s fine to do that for now. Though the types like RunPolicy is used in MPI operator so that should be reused.

**Comment by gaocegege** [comment]
Yeah, I think so. 

BTW @YesterdayxD @4everming

Really appreciate it if you could open an issue to illustrate why we need the change and how to implement it in several operators.

Personally, I know that there is a queue in PodGroup and we can set it to define the priority. And the default value is `default`. But I think we should have a doc or issue to keep track of the change. Maybe we could cc k82cn@ to request a review, too.

**Comment by YesterdayxD** [comment]
ok，I will open an issue about the PR

**Comment by k82cn** [comment]
that's great, I'd like to contribute too :)

**Comment by 4everming** [comment]
> Yeah, I think so.
> 
> BTW @YesterdayxD @4everming
> 
> Really appreciate it if you could open an issue to illustrate why we need the change and how to implement it in several operators.
> 
> Personally, I know that there is a queue in PodGroup and we can set it to define the priority. And the default value is `default`. But I think we should have a doc or issue to keep track of the change. Maybe we could cc k82cn@ to request a review, too.

Sure thing. Thanks a lot.

**Comment by gaocegege** [comment]
https://github.com/kubeflow/common/issues/46

@4everming and @YesterdayxD are working on a proposal, @ me and k82cn@ if you think it is ready to review.

**Comment by terrytangyuan** [comment]
Closing this due to inactivity. Feel free to re-open if this is picked up again.

**Pull request closed without merging**